shakenfist/db.py

```diff
@@ -146,13 +161,86 @@ def persist_ipmanager(network_uuid, data):
 def delete_ipmanager(network_uuid):
     etcd.delete('ipmanager', None, uuid)
 
+#####################################################################
+# Instances
+#####################################################################
+
+# Theorem: we never write to the instance object, except for when we
+# create or delete it. Anything which changes value during the lifetime
+# of the instance (state, power_state, etc) is stored in a set of
+# secondary values in etcd that we never cache. This is version 2 of
+# the instance object.
+
+
+def create_instance(instance_uuid, cpus, disk_spec, memory, name, namespace,
+                    requested_placement, ssh_key, user_data, video):
+    metadata = {
+        'uuid': instance_uuid,
+
+        'cpus': cpus,
+        'disk_spec': disk_spec,
+        'memory': memory,
+        'name': name,
+        'namespace': namespace,
+        'requested_placement': requested_placement,
+        'ssh_key': ssh_key,
+        'user_data': user_data,
+        'video': video,
+
+        'version': 2
+    }
+    etcd.put('instance', None, instance_uuid, metadata)
+
 
 def get_instance(instance_uuid):
-    return etcd.get('instance', None, instance_uuid)
+    # TODO(mikal): handle upgrade here
```

Comment:
I was going to suggest a separate upgrade tool but having the ability to auto-upgrade without downtime is pretty cool. 

I don't like the complexity of having upgrade code in the middle of the main code, especially because it will keep growing.

I don't have a good solution if we want to keep the ability to upgrade without downtime.

Reply:
Its a fair point. I think if we go this route we will _have_ to do some upgrade CI as well to ensure that existing instances aren't clobbered. An upgrade tool is tempting as I agree it would make this code a lot simpler.